src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DocumentPresentation/AbstractTextDocumentPresentationEndpointBase.cs

```diff
@@ -272,7 +273,7 @@ private TextDocumentEdit[] MapDocumentChanges(TextDocumentEdit[] documentEdits,
                     TextDocument = new OptionalVersionedTextDocumentIdentifier()
                     {
                         Uri = razorDocumentUri,
-                        Version = entry.TextDocument.Version
+                        Version = hostDocumentVersion
```

Comment:
Problem number 2: I was using the projected document version, not the host document version.

(In my defense for this one, when I wrote this code the web tools side of things wasn't merged, so I couldn't do any end-to-end testing)